libs/geo/src/main/java/org/elasticsearch/geo/geometry/GeometryVisitor.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.geo.geometry;
+
+/**
+ * Support class for creating of geometry Visitors
+ */
+public interface GeometryVisitor<T> {
+
+    T visit(Circle circle);
+
+    T visit(GeometryCollection<?> collection);
+
+    T visit(Line line);
+
+    T visit(LinearRing ring);
+
+    T visit(MultiLine multiLine);
+
+    T visit(MultiPoint multiPoint);
+
+    T visit(MultiPolygon multiPolygon);
+
+    T visit(Point point);
+
+    T visit(Polygon polygon);
+
+    T visit(Rectangle rectangle);
+
+}
```

Comment:
I don't really understand why we have this class. There are no test for it and not sure if needed. Could you justify it?

Reply:
This is an implementation of the [Visitor](https://en.wikipedia.org/wiki/Visitor_pattern) pattern. The basic idea is to simplify adding new operations on Geometries, without constantly modifying and adding new functionality to the Geometry hierarchy and keeping it as lightweight as possible. 

As an example, I used it for [WKT serialization](https://github.com/imotov/elasticsearch/blob/9b784e24520d385ceac533673be28a9fc4c4650d/libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java#L71). Later I can add XContent serialization in another project. The alternative to this approach is to have a large 

```
if (obj instanceof This) { 
  doThis((This) obj); 
} elseif (obj instanceof That) {
  doThat((That) obj);
...
} else {
  throw new IllegalArgumentException("Unknown object " + obj);
}
```
The Visitor Pattern replaces this structure with Interface inheritance making it easier to identify all places that are using this structure, and making a shape a compile-time failure instead of runtime. I think it is a cleaner approach. This class is tested as part of toWKT code path, but you are right, I should add a designated test for it.